ml/utils.py

```diff
@@ -0,0 +1,354 @@
+    _, indices_ = nn.kneighbors(transformed_values[random_points])
+    indices = np.unique(indices_.squeeze())
+
+    f, ax = plt.subplots()
+    ax.scatter(x=two_dim_data[indices, 0], y=two_dim_data[indices, 1])
+    # adding the legend
+    for x in indices:
+        ax.text(
+            x=two_dim_data[x, 0], y=two_dim_data[x, 1], s=sorted_values[x], fontsize=8
+        )
+    ax.set_title(
+        "multi-dimensional-scaling representation using a 3gram similarity matrix"
+    )
+
+    f2, ax2 = plt.subplots(figsize=(7, 7))
+    cax2 = ax2.matshow(transformed_values[indices, :][:, indices])
+    ax2.set_yticks(np.arange(len(indices)))
+    ax2.set_xticks(np.arange(len(indices)))
+    ax2.set_yticklabels(sorted_values[indices], rotation="30")
+    ax2.set_xticklabels(sorted_values[indices], rotation="60", ha="right")
+    ax2.xaxis.tick_bottom()
+    ax2.set_title("Similarities across categories")
+    f2.colorbar(cax2)
+    f2.tight_layout()
+
+
+# #################################################################################################
+#
+#   Finding subgraphs within a large graph.
+#
+# #################################################################################################
+
+
+def search_to_df(word, col, df):
+    """
+    A simple way to retrieve entries from a given col in a dataframe
+    :eg
+        search_to_df('BlackRock', 'to_node', edf)
+
+    :param word: str search term
+    :param col: given column of dataframe
+    :param df: pandas dataframe
+    :returns
+        DataFrame of results
+    """
+    word = word.lower()
+    try:
+        res = df[df[col].apply(lambda x: word in str(x).lower())]
+    except TypeError as e:
+        logger.error(e)
+        return df
+    return res
+
+
+def get_nearest(search_term, src, dst, edf):
+    """
+    :param search_term:
+    :param src:
+    :param dst:
+    :param edf:
+    :return: pandas.DataFrame
+    """
+    logger.info(f"Finding {search_term} in both {src} and {dst} columns")
+    tdf = pd.concat(
+        [search_to_df(search_term, src, edf), search_to_df(search_term, dst, edf)],
+        axis=0,
+    )
+    return tdf
+
+
+def get_graphistry_from_search(search_term, src, dst, node_col, edf, ndf):
+    """
+        Useful helper function to get subgraph from a search term
+    :param search_term: Note this retrieves all nodes that have `search_term` in it -- ie, not strict matches
+    :param src:
+    :param dst:
+    :param node_col:
+    :param edf:
+    :param ndf:
+    :return:
+    """
+    tdf = get_nearest(search_term, src, dst, edf)
+    gcols = pd.concat(
+        [tdf[dst], tdf[src]], axis=0
+    )  # get all node entries that show up in edge graph
+    ntdf = ndf[ndf[node_col].isin(gcols)]
+    g = graphistry.edges(tdf, src, dst).nodes(ntdf, node_col)
+    return g
+
+
+# lets make simple functions to find subgraphs by search_term = a given node
+def get_milieu_graph_from_search(search_term, src, dst, edf, both=False):
```

Comment:
is milieu a technical term? is it different from neighborhood?

Reply:
:) `get_k_hop(k=2)` and exn if `k != 2` :)

Reply:
Also, cugraph/networkx seem to have already: https://docs.rapids.ai/api/cugraph/stable/api_docs/api/cugraph.community.egonet.ego_graph.html?highlight=neighbors#cugraph-community-egonet-ego-graph